src/Sylius/Bundle/ApiBundle/DataProvider/PaymentMethodsCollectionDataProvider.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\DataProvider;
+
+use ApiPlatform\Core\DataProvider\ContextAwareCollectionDataProviderInterface;
+use ApiPlatform\Core\DataProvider\RestrictedDataProviderInterface;
+use Sylius\Component\Channel\Context\ChannelContextInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+use Sylius\Component\Core\Repository\OrderRepositoryInterface;
+use Sylius\Component\Core\Repository\PaymentMethodRepositoryInterface;
+use Sylius\Component\Core\Repository\PaymentRepositoryInterface;
+use Sylius\Component\Payment\Resolver\PaymentMethodsResolverInterface;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class PaymentMethodsCollectionDataProvider implements ContextAwareCollectionDataProviderInterface, RestrictedDataProviderInterface
+{
+    public function __construct(
+        private OrderRepositoryInterface $orderRepository,
+        private PaymentMethodRepositoryInterface $paymentMethodRepository,
+        private PaymentRepositoryInterface $paymentRepository,
+        private ChannelContextInterface $channelContext,
+        private PaymentMethodsResolverInterface $paymentMethodsResolver
+    ) {
+    }
+
+    public function getCollection(string $resourceClass, string $operationName = null, array $context = [])
+    {
+        /** @var ChannelInterface $channel */
+        $channel = $this->channelContext->getChannel();
+
+        if (!isset($context['filters'])) {
+            return $this->paymentMethodRepository->findEnabledForChannel($channel);
+        }
+
+        $parameters = $context['filters'];
+
+        if (!array_key_exists('tokenValue', $parameters)) {
+           return [];
+        }
+
+        if (!array_key_exists('paymentId', $parameters)) {
+           return [];
+        }
```

Comment:
Still clear enough imo:
```suggestion
        if (!array_key_exists('tokenValue', $parameters) || !array_key_exists('paymentId', $parameters)) {
           return [];
        }
```